File: app/src/androidTest/java/org/kiwix/kiwixmobile/settings/KiwixSettingsFragmentTest.java
/*
 * Kiwix Android
 * Copyright (c) 2019 Kiwix <android.kiwix.org>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */

package org.kiwix.kiwixmobile.settings;

import android.Manifest;
import android.view.View;
import androidx.annotation.StringRes;
import androidx.test.rule.ActivityTestRule;
import androidx.test.rule.GrantPermissionRule;
import com.schibsted.spain.barista.interaction.BaristaSleepInteractions;
import org.hamcrest.Matcher;
import org.jetbrains.annotations.NotNull;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.kiwix.kiwixmobile.core.R;
import org.kiwix.kiwixmobile.main.KiwixMainActivity;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.contrib.RecyclerViewActions.actionOnItem;
import static androidx.test.espresso.matcher.ViewMatchers.hasDescendant;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static com.schibsted.spain.barista.assertion.BaristaVisibilityAssertions.assertDisplayed;
import static org.hamcrest.Matchers.anyOf;
import static org.kiwix.kiwixmobile.testutils.TestUtils.TEST_PAUSE_MS;
import static org.kiwix.kiwixmobile.utils.StandardActions.enterSettings;
import static org.kiwix.kiwixmobile.utils.StandardActions.openDrawer;

public class KiwixSettingsFragmentTest {
  @Rule
  public ActivityTestRule<KiwixMainActivity> activityTestRule =
    new ActivityTestRule<>(KiwixMainActivity.class);
  @Rule
  public GrantPermissionRule readPermissionRule =
    GrantPermissionRule.grant(Manifest.permission.READ_EXTERNAL_STORAGE);
  @Rule
  public GrantPermissionRule writePermissionRule =
    GrantPermissionRule.grant(Manifest.permission.WRITE_EXTERNAL_STORAGE);

  @Before
  public void setup() {
    BaristaSleepInteractions.sleep(TEST_PAUSE_MS);
    openDrawer();
    enterSettings();
  }

  @Test
  public void testToggle() {
    clickOn(R.string.pref_back_to_top);
    clickOn(R.string.pref_newtab_background_title);
    clickOn(R.string.pref_external_link_popup_title);
    clickOn(R.string.pref_wifi_only);
  }

  private void clickOn(@StringRes int... stringIds) {
    Matcher<View>[] matchers = new Matcher[stringIds.length];
    for (int i = 0; i < stringIds.length; i++) {
      matchers[i] = withText(stringIds[i]);
    }
    onView(withId(R.id.recycler_view))
      .perform(actionOnItem(hasDescendant(anyOf(matchers)), click()));
  }

  @Test
  public void testLanguageDialog() {
    clickOn(R.string.device_default);
    assertDisplayed(R.string.pref_language_title);
  }

  @Test
  public void testStorageDialog() {
    clickOn(R.string.internal_storage, R.string.external_storage);
    assertDisplayed(R.string.pref_storage);
  }

  @Test
  public void testHistoryDialog() {
    clickOn(R.string.pref_clear_all_history_title);
    assertDisplayed(R.string.clear_all_history_dialog_title);
  }

  @Test
  public void testNightModeDialog() {
    clickOn(R.string.pref_night_mode);
    for (String nightModeString : nightModeStrings()) {
      assertDisplayed(nightModeString);
    }
  }

  @NotNull private String[] nightModeStrings() {
    return activityTestRule.getActivity()
      .getResources()
      .getStringArray(R.array.pref_night_modes_entries);
  }
}

File: app/src/androidTest/java/org/kiwix/kiwixmobile/settings/KiwixSettingsFragmentTest.kt
/*
 * Kiwix Android
 * Copyright (c) 2019 Kiwix <android.kiwix.org>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */
package org.kiwix.kiwixmobile.settings

import androidx.compose.ui.test.junit4.accessibility.enableAccessibilityChecks
import androidx.compose.ui.test.junit4.createComposeRule
import leakcanary.LeakAssertions
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.kiwix.kiwixmobile.BaseActivityTest
import org.kiwix.kiwixmobile.core.main.CoreMainActivity
import org.kiwix.kiwixmobile.core.utils.TestingUtils.COMPOSE_TEST_RULE_ORDER
import org.kiwix.kiwixmobile.core.utils.TestingUtils.RETRY_RULE_ORDER
import org.kiwix.kiwixmobile.splash.splash
import org.kiwix.kiwixmobile.main.KiwixMainActivity
import org.kiwix.kiwixmobile.testutils.RetryRule
import org.kiwix.kiwixmobile.ui.KiwixDestination
import org.kiwix.kiwixmobile.utils.StandardActions

class KiwixSettingsFragmentTest : BaseActivityTest() {
  @Rule(order = RETRY_RULE_ORDER)
  @JvmField
  val retryRule = RetryRule()

  @get:Rule(order = COMPOSE_TEST_RULE_ORDER)
  val composeTestRule = createComposeRule()

  lateinit var kiwixMainActivity: KiwixMainActivity

  @Before
  override fun waitForIdle() {
    super.waitForIdle()
    launchMainActivity()
    composeTestRule.enableAccessibilityChecks(createAccessibilityValidator())
    // Go to IntroScreen
    activityScenario.onActivity {
      kiwixMainActivity = it
      it.navigate(KiwixDestination.Intro.route)
    }
    composeTestRule.waitForIdle()
    splash {
      swipeLeft(composeTestRule)
      clickGetStarted(composeTestRule) {}
    }
    StandardActions.openDrawer(kiwixMainActivity as CoreMainActivity)
    StandardActions.enterSettings(composeTestRule)
  }

  @Test
  fun testSettingsActivity() {
    settingsRobo {
      assertZoomTextViewPresent(composeTestRule)
      clickNightModePreference(composeTestRule)
      assertNightModeDialogDisplayed(composeTestRule)
      dismissDialog()
      toggleBackToTopPref(composeTestRule)
      toggleOpenNewTabInBackground(composeTestRule)
      toggleExternalLinkWarningPref(composeTestRule)
      toggleWifiDownloadsOnlyPref(composeTestRule)
      clickExternalStoragePreference(composeTestRule)
      assertExternalStorageSelected(composeTestRule)
      clickInternalStoragePreference(composeTestRule)
      assertInternalStorageSelected(composeTestRule)
      clickClearHistoryPreference(composeTestRule)
      assertHistoryDialogDisplayed(composeTestRule)
      dismissDialog()
      clickClearNotesPreference(composeTestRule)
      assertNotesDialogDisplayed(composeTestRule)
      dismissDialog()
      clickExportBookmarkPreference(composeTestRule)
      assertExportBookmarkDialogDisplayed(composeTestRule)
      dismissDialog()
      clickOnImportBookmarkPreference(composeTestRule)
      assertImportBookmarkDialogDisplayed(composeTestRule)
      dismissDialog()
      clickLanguagePreference(composeTestRule, activityScenario)
      assertLanguagePrefDialogDisplayed(composeTestRule, activityScenario)
      dismissDialog()
      assertVersionTextViewPresent(composeTestRule)
      clickCredits(composeTestRule)
      assertContributorsDialogDisplayed(composeTestRule)
      dismissDialog()
    }
    LeakAssertions.assertNoLeaks()
  }
}
